zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Observer.java

```diff
@@ -114,11 +115,17 @@ void observeLeader() throws Exception {
                     throw new Exception("learned about role change");
                 }
 
-                self.setLeaderAddressAndId(master.addr, master.getId());
-                self.setZabState(QuorumPeer.ZabState.SYNCHRONIZATION);
-                syncWithLeader(newLeaderZxid);
-                self.setZabState(QuorumPeer.ZabState.BROADCAST);
-                completedSync = true;
+                final long startTime = Time.currentElapsedTime();
+                try {
+                    self.setLeaderAddressAndId(master.addr, master.getId());
+                    self.setZabState(QuorumPeer.ZabState.SYNCHRONIZATION);
+                    syncWithLeader(newLeaderZxid);
+                    self.setZabState(QuorumPeer.ZabState.BROADCAST);
+                    completedSync = true;
+                } finally {
+                    final long syncTime = Time.currentElapsedTime() - startTime;
+                    ServerMetrics.getMetrics().OBSERVER_SYNC_TIME.add(syncTime);
```

Comment:
I think we should calculate this time iff `completedSync` is true. Otherwise, we will get noisy data such as 0 sync time in cases where sync immediately failed (due to network partition for example). 

For the same reason the existing follower sync time needs to apply the same fix, which can be done separately.

Reply:
Agreed. Changed. 

I have also opened a JIRA for fix the follower sync time and will submit the fix shortly. https://issues.apache.org/jira/browse/ZOOKEEPER-4318

Reply:
@hanm thanks for reviewing and approving the follower_sync_time. Can you approve this one, so the change can be merged?